ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/serde/json/ValueSpecJsonSchemaSerdeSupplier.java

```diff
@@ -0,0 +1,102 @@
+ */
+
+package io.confluent.ksql.test.serde.json;
+
+import io.confluent.connect.json.JsonSchemaConverter;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.json.JsonSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.serde.SerdeFeatures;
+import io.confluent.ksql.serde.SerdeUtils;
+import io.confluent.ksql.serde.connect.ConnectSchemas;
+import io.confluent.ksql.serde.json.JsonSchemaTranslator;
+import io.confluent.ksql.test.serde.ConnectSerdeSupplier;
+import java.util.Map;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.connect.data.Schema;
+
+public class ValueSpecJsonSchemaSerdeSupplier extends ConnectSerdeSupplier<JsonSchema> {
+  private final JsonSchemaTranslator schemaTranslator;
+  private final LogicalSchema logicalSpecSchema;
+  private final SerdeFeatures serdeFeatures;
+
+  public ValueSpecJsonSchemaSerdeSupplier(
+      final LogicalSchema logicalSpecSchema,
+      final SerdeFeatures serdeFeatures
+  ) {
+    super(JsonSchemaConverter::new);
+    this.schemaTranslator = new JsonSchemaTranslator();
+    this.logicalSpecSchema = logicalSpecSchema;
+    this.serdeFeatures = serdeFeatures;
+  }
+
+  @Override
+  protected Schema fromParsedSchema(final JsonSchema schema) {
+    return schemaTranslator.toConnectSchema(schema);
+  }
+
+  @Override
+  public Deserializer<Object> getDeserializer(
+      final SchemaRegistryClient schemaRegistryClient,
+      final boolean isKey
+  ) {
+    final Schema outerSchema = ConnectSchemas.columnsToConnectSchema(
+        isKey ? logicalSpecSchema.key() : logicalSpecSchema.value());
+
+    final Schema innerSchema = SerdeUtils.applySinglesUnwrapping(outerSchema, serdeFeatures);
+
+    return new SpecJsonSchemaDeserializer(
+        super.getDeserializer(schemaRegistryClient, isKey),
+        innerSchema
+    );
+  }
+
+  private static class SpecJsonSchemaDeserializer implements Deserializer<Object> {
+    private final Deserializer<Object> internalDeserializer;
+    private final Schema specSchema;
+
+    SpecJsonSchemaDeserializer(
+        final Deserializer<Object> internalDeserializer,
+        final Schema specSchema
+    ) {
+      this.internalDeserializer = internalDeserializer;
+      this.specSchema = specSchema;
+    }
+
+    @Override
+    public Object deserialize(final String topic, final byte[] bytes) {
+      final Object deserialized = internalDeserializer.deserialize(topic, bytes);
+      if (!(deserialized instanceof Map)) {
+        return deserialized;
+      }
+
+      final Map<String, Object> deserializedMap = (Map<String, Object>) deserialized;
+
+      // This makes JSON_SR output compatible with old QTT tests that used to deserialize
+      // JSON_SR with a JSON object mapper instead of the ConnectSerdeSupplier.
+      // The ConnectSerdeSupplier is adding the missing fields with null values even if they're
```

Comment:
QQ: why the field is serialized to map if it's not in schema? I might miss something...

Reply:
This is what the ConnectSerdeSupplier is returning. I got lost on this part for some time. After debugging I found out that QTT creates a Map with the columns and its deserialized values. This is the part that is special for QTT.